src/main/java/org/jmxtrans/agent/Query.java

```diff
@@ -131,20 +131,35 @@ public void collectAndExport(@Nonnull MBeanServer mbeanServer, @Nonnull OutputWr
 
         for (ObjectName on : objectNames) {
             try {
-                if (attribute == null || attribute.isEmpty()) {
-                    for (MBeanAttributeInfo mBeanAttributeInfo : mbeanServer.getMBeanInfo(on).getAttributes()) {
-                        collectAndExportAttribute(mbeanServer, outputWriter, on, mBeanAttributeInfo.getName());
-                    }
-
-                } else {
-                    collectAndExportAttribute(mbeanServer, outputWriter, on, attribute);
-                }
+                collectAndExportForObjectName(mbeanServer, on, outputWriter);
             } catch (Exception e) {
                 logger.log(Level.WARNING, "Exception collecting " + on + "#" + attribute + (key == null ? "" : "#" + key), e);
             }
         }
     }
 
+    private void collectAndExportForObjectName(MBeanServer mbeanServer, ObjectName on, OutputWriter outputWriter)
+            throws Exception {
+        for (String attribute : resolveAttributes(mbeanServer, on)) {
```

Comment:
I am not sure it is a good idea to cache the attributes in case empty/null attribute configuration is supplied. In that case, all attributes are fetched from the MBean. Caching the attributes would prevent us from collecting attributes added to the MBean at a later point (I think that can happen, correct me if I'm wrong)

We could cache the attributes in case of a comma-separated list, but I would argue the minimal performance benefit is not worth the extra complexity and state required.